### PR TITLE
Consider the OperationSegment when determining the TargetStructuredType

### PR DESCRIPTION
**Pull request by ificator**
### Issues

Fixes https://github.com/OData/WebApi/issues/1610

### Description

If you have an auto-expanded property on an object:
```
public class A
{
  [AutoExpand]
  public B B { get; set; }
}
```
And a function `F` bound to `A` that returns type `C`, the auto-expand logic incorrectly attempts to expand `B` on `C` if a (any) query string is provided:
```
GET .../A/F?a=b

The query specified in the URI is not valid. Could not find a property named 'B' on type 'C'.
```
Since the fix appeared to be pretty straight forward (make sure `OperationSegment` is handled in the path when trying to determine the `TargetStructuredType`) I've decided to push this PR.

NOTE: I'm also fixing a couple of bonus issues
1. There's a bug whereby the existing `EnableQuery_Works_WithAutoExpand` test wasn't validating auto-expand because `EnableQueryCategory` was not being registered as an `EntityType`.
2. VisualStudio complained about the string interpolation in `EdmTypeCacheItemComparer.GetHashCode`, but methods used for comparisons (e.g. `Equals` and `GetHashCode`) should NOT allocate so I rewrote it to use a hash combining algorithm instead.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### CODE REVIEW

File: src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs
Comment (ificator) [inline src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs:77]: Or can we trust it? the `Equals` implementation assumes that `Equals` is implemented, which it isn't, so that would suggest we're dealing with instance equality...

File: src/Microsoft.AspNet.OData.Shared/Formatter/EdmLibHelpers.cs
Comment (ificator) [inline src/Microsoft.AspNet.OData.Shared/Formatter/EdmLibHelpers.cs:662]: Shouldn't be limited to functions

Comment (ificator) [reply src/Microsoft.AspNet.OData.Shared/Formatter/EdmLibHelpers.cs:662]: I've fixed this locally, but I'll hold off pushing a new iteration until we've resolved my previous question re: `GetHashCode`

File: src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs
Comment (ificator) [reply src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs]: I've convinced myself that the default `GetHashCode` implementation will work for us.

**Comment by ificator** [comment]
@KenitoInc / @gathogojr could you take a look at this PR?

File: src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs
Comment (gathogojr) [inline src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs:78]: @ificator I'm not sure I understand the logic here... What value does line 77 add if the assigned value is overwritten here?

**Comment by gathogojr** [comment]
Hi @ificator. Thank you for your contribution. Looks great. I agree with you that ideally methods used for comparisons should not allocate but I've left some comments around the changes you introduced to `GetHashCode` method.

File: src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs
Comment (gathogojr) [reply src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs:78]: argh, never mind. I just realized that the value is not exactly being overwritten

**Review by gathogojr — APPROVED** [review]
LGTM

**Review by KenitoInc — APPROVED** [review]

File: src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs
Comment (ificator) [reply src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs:78]: Yeah, this is a somewhat standard algorithm for generating hashcodes that's taken from "Effective Java" (https://stackoverflow.com/questions/11537005/why-chose-31-to-do-the-multiplication-in-the-hashcode-implementation). I became less concerned with my change once I found the unit tests for the cache and they were all passing.

**Review by mikepizzo — APPROVED** [review]
:shipit:

**Comment by ificator** [comment]
So who gets to push the merge button? 😁

**Review by xuzhg — APPROVED** [review]
:shipit:

**Comment by ificator** [comment]
Thanks everyone!

**Comment by mikepizzo** [comment]
Thank you Brad!

**Pull request merged**
